File: lib/java/openlinktoken/src/main/java/org/openlinktoken/tokens/InferenceSignatureProvider.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokens;

import org.openlinktoken.attributes.Attribute;
import java.util.List;
import java.util.Map;

/**
 * SPI for inference-based token signature generation.
 *
 * <p>Implementations are discovered at runtime via {@link java.util.ServiceLoader}.
 * When no implementation is on the classpath, inference-based tokens are silently
 * disabled and only standard attribute-expression tokens (T1–T5) are generated.
 *
 * <p>Each provider is responsible for one token identifier (e.g. "ML1") and reports
 * it via {@link #getTokenId()}.
 */
public interface InferenceSignatureProvider {

    /**
     * Return the token identifier this provider handles (e.g. {@code "ML1"}).
     */
    String getTokenId();

    /**
     * Return whether this provider is currently enabled and configured.
     */
    boolean isEnabled();

    /**
     * Generate a single inference-based token signature from the given person attributes.
     *
     * @param personAttributes normalised attribute map for one record
     * @return hex-encoded signature string, or {@code null} if the record is invalid
     */
    String generateSignature(Map<Class<? extends Attribute>, String> personAttributes);

    /**
     * Generate inference-based token signatures for a batch of records in a single
     * inference pass, also returning the raw embedding vectors needed for rotation tokens.
     *
     * @param rows list of normalised attribute maps, one per record
     * @return {@link InferenceBatchResult} with parallel lists of signatures and embeddings
     */
    InferenceBatchResult generateBatch(List<Map<Class<? extends Attribute>, String>> rows);
}

File: lib/python/openlinktoken/src/main/openlinktoken/tokens/inference_signature_provider.py
# SPDX-License-Identifier: MIT

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Protocol, Type, runtime_checkable

from openlinktoken.attributes.attribute import Attribute


@dataclass
class InferenceBatchResult:
    """Results of a batched inference pass.

    Attributes:
        signatures: Hex-encoded token signatures in the same order as the input rows.
    """

    signatures: List[str] = field(default_factory=list)


@runtime_checkable
class InferenceSignatureProvider(Protocol):
    """Protocol for inference-based token signature generation.

    Implementations are discovered at runtime via
    ``importlib.metadata.entry_points(group="openlinktoken.inference_providers")``.
    When no implementation is installed, inference-based tokens are silently
    disabled and only standard attribute-expression tokens (T1–T5) are generated.

    Each provider is responsible for one token identifier (e.g. ``"ML1"``) and
    reports it via :meth:`get_token_id`.
    """

    def get_token_id(self) -> str:
        """Return the token identifier this provider handles (e.g. ``"ML1"``)."""
        ...

    def is_enabled(self) -> bool:
        """Return whether this provider is currently enabled and configured."""
        ...

    def generate_signature(self, person_attributes: Dict[Type[Attribute], str]) -> str:
        """Generate a single inference-based token signature.

        Args:
            person_attributes: Normalised attribute map for one record.

        Returns:
            Hex-encoded signature string, or ``None`` if the record is invalid.
        """
        ...

    def generate_batch(self, rows: List[Dict[Type[Attribute], str]]) -> InferenceBatchResult:
        """Generate inference-based signatures for a batch of records in one pass.

        Args:
            rows: List of normalised attribute maps, one per record.

        Returns:
            :class:`InferenceBatchResult` with signatures in input row order.
        """
        ...
